src/client/opamClient.ml

```diff
@@ -1433,14 +1433,10 @@ module PIN = struct
   let unpin st ?(action=true) names =
     let pinned_before = st.pinned in
     let st = unpin st names in
-    let available = Lazy.force st.available_packages in
     let installed_unpinned = (pinned_before -- st.pinned) %% st.installed in
     if action && not (OpamPackage.Set.is_empty installed_unpinned) then
       let atoms =
-        OpamPackage.Set.fold (fun nv acc ->
-            if OpamPackage.Set.mem nv available then
-              (nv.name, Some (`Eq, nv.version)) :: acc
-            else (nv.name, None) :: acc)
+        OpamPackage.Set.fold (fun nv acc -> (nv.name, None) :: acc)
```

Comment:
I think this is the good approach, but I we should probably keep some check on availability (for _any_ version of the package, to remain consistent).

Otherwise, I imagine the following scenario could happen:
- package `foo` was unavailable, but you wanted to try anyway
- so you pinned it to change its constraints and managed to install it
- now you want to unpin it
→ without the check for availability, we would request `upgrade(foo)`, and the solver would fail ("no available package named 'foo'" — leaving you with your pinned package installed until you remove or upgrade all manually!)
→ with a check, ideally, we would remove `foo` from the constraint, and the `upgrade` would naturally remove it

However, we could still encounter the case where the solver fails due to more complex constraints. _Maybe_ this is a good case for a use of `best_effort` instead? It would upgrade the packages if possible, but guarantee consistency anyway by removing them if needed. Could there be unwanted side-effects to this approach ?

Reply:
FWIS, all those checks are already done in the `compute_ugrade` function: as `foo` is an orphan, it will be added to the remove set in the solver request.

Reply:
Ah indeed, `compute_upgrade` has a special handling for that case :+1: